### PR TITLE
Implement functions to encode float/double values according to IEEE-754

### PR DESCRIPTION
**Pull request by ausgoo**

### CODE REVIEW

**Comment by ausgoo** [comment]
@wenleix , could you help to add reviewers in? Thanks.

**Comment by electrum** [comment]
Please add documentation.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java [hunk 1]
```diff
@@ -146,6 +146,26 @@ public static Slice toBigEndian64(@SqlType(StandardTypes.BIGINT) long value)
         return slice;
     }
 
+    @Description("encode value as an IEEE 754 floating-point 'single format' big endian varbinary")
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java]: I understand this comes from the Java doc:
```
Returns a representation of the specified floating-point value
according to the IEEE 754 floating-point "double format" bit layout.
```

However, my understanding is that `IEEE 754 floating-point "double format" bit layout` is a whole thing. And make it to be `IEEE 754 floating-point 'single format' big endian varbinary`.

What about the following (tentative, we will see how others think later :) ):
```
encode value as a big endian varbinary according to IEEE 754 single-precision floating-point format 
```

`single-precision floating-point format` comes from https://en.wikipedia.org/wiki/Single-precision_floating-point_format

Comment (ausgoo) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java]: Sounds good. Will fix it. Thanks.

**Comment by ausgoo** [comment]
@wenleix , all comments have been addressed. Please continue to review. Thanks.

File: presto-docs/src/main/sphinx/functions/binary.rst [hunk 2]
```diff
@@ -38,6 +38,14 @@ Binary Functions
 
     Encodes ``bigint`` in a 64-bit 2's complement big endian format.
 
+.. function:: to_ieee754_binary32(real) -> varbinary
+
+    Encodes ``real`` in a 32-bit binary according to IEEE 754 single-precision floating-point format.
```
Comment (wenleix) [inline presto-docs/src/main/sphinx/functions/binary.rst]: nit: we should also state the bytes ordering is big-endian

**Comment by haozhun** [comment]
Talk with @martint about the name. Add a test each for the two function with numbers that don't show particular patterns in binary representation. It looks good otherwise.

@wenleix will work with you to merge this after addressing my comments. Thank you, @ausgoo

**Comment by ausgoo** [comment]
@haozhun , "Add a test each for the two function with numbers that don't show particular patterns in binary representation" does that mean to number 0.0 and 1.0 in both testcases? I can help to remove these two.

File: presto-docs/src/main/sphinx/functions/binary.rst [hunk 3]
```diff
@@ -38,6 +38,14 @@ Binary Functions
 
     Encodes ``bigint`` in a 64-bit 2's complement big endian format.
 
+.. function:: to_ieee754_binary32(real) -> varbinary
```
Comment (martint) [inline presto-docs/src/main/sphinx/functions/binary.rst]: This should just be called `to_ieee754`

**Comment by ausgoo** [comment]
PR has been updated with the following change:
1. change name to `to_ieee754`.
2. add test for regular numbers in both testcases.

**Comment by wenleix** [comment]
Resolves #8007

**Comment by ausgoo** [comment]
@wenleix , the failure seems to be consistent. Here are the failure stack traces:
```
2017-05-24 06:36:42 INFO Triggering GC to avoid Code Cache eviction bugs
2017-05-24 06:36:45 SEVERE Error processing Split 20170524_005142_01460_jsr94.2.0-9  (start = 890779.283182, wall = 3103 ms, cpu = 0 ms, wait = 0 ms, calls = 2): SUBQUERY_MULTIPLE_ROWS: Scalar sub-query has returned multiple rows
2017-05-24 06:36:45 SEVERE Stage 20170524_005142_01460_jsr94.2 failed
com.facebook.presto.spi.PrestoException: Scalar sub-query has returned multiple rows
	at com.facebook.presto.operator.EnforceSingleRowOperator.addInput(EnforceSingleRowOperator.java:131)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:309)
	at com.facebook.presto.operator.Driver.lambda$processFor$6(Driver.java:234)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:537)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:229)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:623)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:463)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

and 

```
2017-05-24 01:13:22 INFO: FAILURE     /    com.facebook.presto.tests.hive.TestHiveStorageFormats.testInsertIntoPartitionedTable (Groups: storage_formats) took 4.6 seconds
2017-05-24 01:13:22 SEVERE: Failure cause:
com.teradata.tempto.query.QueryExecutionException: java.sql.SQLException: Query failed (#20170524_011318_00152_svr5u): Error committing write to Hive
	at com.teradata.tempto.query.JdbcQueryExecutor.execute(JdbcQueryExecutor.java:124)
	at com.teradata.tempto.query.JdbcQueryExecutor.executeQuery(JdbcQueryExecutor.java:85)
	at com.teradata.tempto.query.QueryExecutor.query(QueryExecutor.java:69)
	at com.facebook.presto.tests.hive.TestHiveStorageFormats.testInsertIntoPartitionedTable(TestHiveStorageFormats.java:171)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.sql.SQLException: Query failed (#20170524_011318_00152_svr5u): Error committing write to Hive
	at com.facebook.presto.jdbc.PrestoResultSet.resultsException(PrestoResultSet.java:1799)
	at com.facebook.presto.jdbc.PrestoResultSet$ResultsPageIterator.computeNext(PrestoResultSet.java:1787)
	at com.facebook.presto.jdbc.PrestoResultSet$ResultsPageIterator.computeNext(PrestoResultSet.java:1756)
	at com.facebook.presto.jdbc.internal.guava.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:145)
	at com.facebook.presto.jdbc.internal.guava.collect.AbstractIterator.hasNext(AbstractIterator.java:140)
	at com.facebook.presto.jdbc.internal.guava.collect.TransformedIterator.hasNext(TransformedIterator.java:42)
	at com.facebook.presto.jdbc.internal.guava.collect.MultitransformedIterator.hasNext(MultitransformedIterator.java:50)
	at com.facebook.presto.jdbc.internal.guava.collect.MultitransformedIterator.hasNext(MultitransformedIterator.java:50)
	at com.facebook.presto.jdbc.PrestoResultSet.next(PrestoResultSet.java:149)
	at com.facebook.presto.jdbc.PrestoStatement.execute(PrestoStatement.java:222)
	at com.facebook.presto.jdbc.PrestoStatement.executeLargeUpdate(PrestoStatement.java:426)
	at com.facebook.presto.jdbc.PrestoStatement.executeUpdate(PrestoStatement.java:398)
	at com.teradata.tempto.query.JdbcQueryExecutor.executeQueryNoParams(JdbcQueryExecutor.java:137)
	at com.teradata.tempto.query.JdbcQueryExecutor.execute(JdbcQueryExecutor.java:117)
	... 16 more
```

Just want to make sure that it is not related to my change, right?

File: presto-docs/src/main/sphinx/functions/binary.rst [hunk 4]
```diff
@@ -38,6 +38,14 @@ Binary Functions
 
     Encodes ``bigint`` in a 64-bit 2's complement big endian format.
 
+.. function:: to_ieee754(real) -> varbinary
+
+    Encodes ``real`` in a 32-bit big-endian binary according to IEEE 754 single-precision floating-point format.
+
+.. function:: to_ieee754(double) -> varbinary
+
+    Encodes ``double`` in a 64-bit big-endian binary according to IEEE 754 single-precision floating-point format.
```
Comment (martint) [inline presto-docs/src/main/sphinx/functions/binary.rst]: Did you mean "double precision"?

Comment (ausgoo) [inline presto-docs/src/main/sphinx/functions/binary.rst]: Yeah, my bad. Will fix it. Thanks.

**Comment by wenleix** [comment]
@ausgoo : Let's add back the test case for 0.0 and 1.0. 

Also, the subject line is a bit too long, ideally, the subject line should be <= 50 chars (https://chris.beams.io/posts/git-commit/)  . 

I looked at https://github.com/wenleix/presto/commit/241bd99ed7a23c314f61b6ccbf3b7c3f57212ed8 and the commit message is simply `Add xxhash64 and to_big_endian_64 functions`, what about

```
Add to_ieee754 function for REAL and DOUBLE
```

?

**Comment by ausgoo** [comment]
@wenleix , sounds good. Will add the tests back and fix the commit messages.

**Comment by ausgoo** [comment]
PR has been updated with the following changes:
1. Fix commit message.
2. Add back 0.0 and 1.0 testcases.

**Comment by wenleix** [comment]
Merged #8091 as https://github.com/prestodb/presto/commit/2819445acd3df6415b2bec64fc5ce88fefefdec0. Thanks!

We had a discussion with @martint  and @haozhun  about the function name. One issue is because how INTEGER/BIGINT will be coerced into REAL or DOUBLE is not clear :). So we decide to have `to_ieee754_32` and `to_ieee754_64` :) .

**Pull request closed without merging**